pkg/binding/interfaces.go

```diff
@@ -49,6 +62,27 @@ type Message interface {
 	Finish(error) error
 }
 
+// ErrNotBinary returned by Message.Binary for non-binary messages.
+var ErrNotBinary = errors.New("message is not in binary mode")
+
+// EventBuilder receives values to build a binary event representation.
+type EventBuilder interface {
+	// Data receives an io.Reader for the data attribute.
+	Data(io.Reader)
+
+	// Set a standard attribute.
+	//
+	// The value can either be the correct type for the attribute, or a canonical
+	// string encoding. See package cloudevents/types
+	Set(attributeId spec.Kind, value interface{})
```

Comment:
The thing I do not like about this is the builder interface does not get to expose the correct types in the api. It is a hidden detail that gets smashed up in a switch as internal details. So you get runtime issues rather that compile errors.

Reply:
This is why I have `SetTime` and `Time` etc on `Event`.

Reply:
We could remove Set() from builder and embed EventContextWriter instead for standard attributes. The idea of using interface{} was to avoid parsing and re-encoding canonical string values if both sides use that format, but its probably not a big enough deal to compromise type safety. I'm fine with using EventContextWriter.